include/aspect/particle/integrator/interface.h

```diff
@@ -0,0 +1,227 @@
+           */
+          virtual ~Interface () {}
+
+          /**
+           * Perform an integration step of moving the particles of one cell
+           * by the specified timestep dt. Implementations of this function
+           * must update the particle location. Between calls to this function
+           * the velocity at the updated particle positions is evaluated and
+           * passed to integrate_step during the next call.
+           *
+           * @param [in] begin_particle An iterator to the first particle to be moved.
+           * @param [in] end_particle An iterator to the last particle to be moved.
+           * @param [in] old_velocities The velocities at t_n, i.e. before the
+           * particle movement.
+           * @param [in] velocities The velocities at t_{n+1}, i.e. after the
+           * particle movement.
+           * @param [in] dt The length of the integration timestep.
+           */
+          virtual
+          void
+          local_integrate_step(const typename std::multimap<LevelInd, Particle<dim> >::iterator &begin_particle,
+                               const typename std::multimap<LevelInd, Particle<dim> >::iterator &end_particle,
+                               const std::vector<Tensor<1,dim> > &old_velocities,
+                               const std::vector<Tensor<1,dim> > &velocities,
+                               const double dt) = 0;
+
+          /**
+           * This function is called at the end of every integration step.
+           * In case of multi-step integrators is signals the beginning of a
+           * new integration step.
+           */
+          virtual void advance_step();
+
+          /**
+           * @return This function returns true if another integration step is required
+           * by the integrator. Its default implementation returns false.
+           */
+          virtual bool continue_integration() const;
+
+          /**
+           * Return data length of the integration related data required for
+           * communication in terms of number of doubles.
+           *
+           * @return The number of doubles required to store the relevant
+           * integrator data for one particle.
+           */
+          virtual unsigned int data_length() const;
+
+          /**
+           * Read integration related data for a particle specified by id_num
+           * from the data array.
+           *
+           * @param [in] data The array of double data to read from.
+           * @param [in] id_num The id number of the particle to read the data
+           * for.
+           */
+          virtual void read_data(const void *&data,
+                                 const particle_index id_num);
+
+          /**
+           * Write integration related data to a vector for a particle
+           * specified by id_num.
+           *
+           * @param [in,out] data The vector of doubles to write integrator
+           * data into.
```

Comment:
This suggests that you advance the pointer (so you take it as a reference). Can you state a postcondition by how many bytes/doubles the pointer is advanced by this function?